### ISSUE TITLE
Feature Request: `StreamingKFoldSplitter` for Efficient CV and Train/Test Splits in Hugging Face Datasets

### ISSUE THREAD

**Issue by dahlem** [open-disease-risk/sat#28]
### 📌 Feature Request: `StreamingKFoldSplitter` for Efficient CV and Train/Test Splits in Hugging Face Datasets

**Summary:**  
Add a utility class `StreamingKFoldSplitter` to support efficient and reproducible train/val/test splits for Hugging Face datasets using streaming mode. This is especially useful when working with large datasets that don’t fit in memory and when deterministic, reusable fold splits are needed.

---

### ✨ Motivation

Currently, there's no built-in utility for:

- Performing **k-fold cross-validation** on large datasets in streaming mode.
- Generating **reusable train/test splits** without saving massive intermediate files.
- Seamlessly integrating existing dataset splits (`train`, `validation`, `test`) when available.

This utility would support workflows that require:
- Scalable preprocessing and model evaluation.
- Reproducible results across distributed environments.
- Flexible switching between one-off and cross-validation strategies.

---

### ✅ Proposed Solution

Implement a class `StreamingKFoldSplitter` with the following features:

#### Class: `StreamingKFoldSplitter`

```python
StreamingKFoldSplitter(
    dataset_name: str,
    id_field: str = "id",
    k: Optional[int] = None,
    val_ratio: float = 0.2,
    test_ratio: float = 0.1,
    test_split_strategy: str = "hash",  # or "existing"
    split_names: Tuple[str, ...] = ("train", "validation", "test")
)
```

#### Method:
```python
load_streaming_split(fold_index: Optional[int] = None) -> Tuple[Iterable, Optional[Iterable], Iterable]
```

Returns:
- `train_dataset` (streaming iterable)
- `val_dataset` (optional, only if fold_index or val_ratio is used)
- `test_dataset` (always returned)

---

### 🔍 Implementation Notes

- Hash-based splitting is performed using MD5 on the specified `id_field`.
- If `test_split_strategy="hash"`: split off a fixed proportion for testing.
- If `test_split_strategy="existing"`: use built-in test set; merge `train` and `validation` for k-fold CV.
- If `k=None`: use `val_ratio` to create a one-shot train/val split.
- Fully compatible with Hugging Face’s `load_dataset(..., streaming=True)` API.
- Does **not** require materializing the full dataset in memory.

---

### 📦 Example Usage

```python
# One-shot split
splitter = StreamingKFoldSplitter("ag_news", id_field="text", k=None, val_ratio=0.2)
train, val, test = splitter.load_streaming_split()

# 5-fold cross-validation
splitter = StreamingKFoldSplitter("ag_news", id_field="text", k=5)
train, val, test = splitter.load_streaming_split(fold_index=1)

# Using existing train/val/test splits
splitter = StreamingKFoldSplitter(
    "glue", id_field="sentence1", k=5,
    test_split_strategy="existing",
    split_names=("train", "validation", "test")
)
train, val, test = splitter.load_streaming_split(fold_index=3)
```

---

### 🛠️ Future Extensions

- Support non-streaming datasets with `.select()`-based filtering.
- Add caching or Parquet export for persistent fold extraction.
- PyTorch and TensorFlow `DataLoader` integration.
- CLI or script interface for fold metadata generation.

---

### 🙌 Contribution Notes

This is a self-contained feature and could live under a utility module like `utils/data_splits.py` or be released as a standalone helper script. Happy to help design tests or integrate into existing workflows!
